File: Grind75/37-ProductExceptSelf/cpp/ProductExceptSelf.cpp
#include "ProductExceptSelf.h"
#include <vector>
#include <prettyPrint.h>

using namespace std;
 
vector<int> ProductExceptSelf::productExceptSelf(vector<int>& nums) {
  size_t N = nums.size();
  if (N < 2)
    return {};
  vector<int> productsForward(N), productsBackward(N), productsExceptSelf(N);

  productsForward[0] = nums[0];
  productsBackward[N-1] = nums[N-1];
  for (size_t i = 1, j = N-2; i < N && j >= 0; i++, j--) {
    productsForward[i] = productsForward[i-1] * nums[i];
    productsBackward[j] = productsBackward[j+1] * nums[j];
  }

  productsExceptSelf[0] = productsBackward[1];
  productsExceptSelf[N-1] = productsForward[N-2];
  for (size_t i = 1; i < N-1; i++) {
    productsExceptSelf[i] = productsForward[i-1] * productsBackward[i+1];
  }

  return productsExceptSelf;
}

File: Grind75/37-ProductExceptSelf/python/ProductExceptSelf.py
#!/bin/python3

from __future__ import annotations
import sys

sys.path.append('../../Utility/python')
from typing import List
from PrettyPrint import colorString
from Colors import *

class Solution(object):
  def productExceptSelf(self, nums: List[int]) -> List[int]:
    N: int = len(nums)
    if N < 2:
      return []
    prodExceptSelf: List[int] = [ 1 ] * N
    for i in range(1, N, 1):
      prodExceptSelf[i] = prodExceptSelf[i-1] * nums[i-1]
    prodsBackward = 1
    for i in range(N-1, -1, -1):
      prodExceptSelf[i] *= prodsBackward
      prodsBackward *= nums[i]
    return prodExceptSelf

if __name__ == '__main__':
  soln = Solution()
  testCases = [
    ([], []),
    ([1], []),
    ([1,2], [2,1]),
    ([1,2,3], [6,3,2]),
    ([-1,2,3], [6,-3,-2]),
    ([0,2,3], [6,0,0]),
    ([3,2,4,-1,-5], [40,60,30,-120,-24]),
  ]

  for nums, answer in testCases:
    print(colorString("test:", LIGHT_BLUE), nums)
    print("       =>", answer)
    result = soln.productExceptSelf(nums)
    print(colorString("       ->", LIGHT_YELLOW), result, end=' ')
    if result == answer:
      print(colorString("PASS", GREEN))
    else:
      print(colorString("FAIL", RED))

